packages/pubsub/src/subscription.js

```diff
@@ -592,30 +365,249 @@ Subscription.prototype.decorateMessage_ = function(message) {
+  this.messages_.add(message);
+  return message;
+};
+
+/**
+ *
+ */
+Subscription.prototype.modifyAckDeadline_ = function(messages, callback) {
+  var self = this;
+  var reqOpts;
+
+  if (this.connection) {
+    reqOpts = messages.reduce(function(reqOpts, message) {
+      message.modifyDeadlineAckIds.push(message.ackId);
+      message.modifyDeadlineSeconds.push(message.deadline);
+      return reqOpts;
+    }, {
+      modifyDeadlineAckIds: [],
+      modifyDeadlineSeconds: []
+    });
+
+    this.connection.write(reqOpts);
+    setImmediate(onComplete);
+    return;
+  }
+
+  var requests = groupByDeadline(messages).map(function(reqOpts) {
+    return self.request({
+      client: 'subscriberClient',
+      method: 'modifyAckDeadline',
+      reqOpts: reqOpts
+    }).catch(function(err) {
+      self.emit('error', err);
+    });
+  });
+
+  Promise.all(requests)
+    .then(onComplete.bind(null, null), onComplete);
+
+  function groupByDeadline(messages) {
+    var requests = new Map();
+
+    messages.forEach(function(message) {
+      if (!requests.has(message.deadline)) {
+        requests.set(message.deadline, {
+          subscription: self.name,
+          ackIds: [],
+          ackDeadlineSeconds: message.deadline
+        });
+      }
+
+      var request = requests.get(message.deadline);
+      request.ackIds.push(message.ackId);
+    });
+
+    return Array.from(requests.values());
+  }
+
+  function onComplete(err, resp) {
+    messages.forEach(function(message) {
+      if (message.deadline === 0) {
+        self.messages_.delete(message);
+      }
+    });
+
+    if (self.connection.isPaused()) {
+      if (self.messages_.size < self.flowControl.maxMessages) {
+        self.connection.resume();
+      }
+    }
+
+    callback(err, resp);
+  }
+};
+
+/**
+ * @param {object} config - The push config.
+ * @param {string} config.pushEndpoint
+ * @param {object} config.attributes
+ */
+Subscription.prototype.modifyPushConfig = function(config, gaxOpts, callback) {
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
+
+  var reqOpts = {
+    subscription: this.name,
+    pushConfig: config
+  };
+
+  this.request({
+    client: 'subscriberClient',
+    method: 'modifyPushConfig',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
+
+/**
+ * @private
+ */
+Subscription.prototype.openConnection_ = function() {
+  var self = this;
+
+  var reqOpts = {
+    subscription: this.name,
+    streamAckDeadlineSeconds: this.ackDeadline
+  };
+
+  this.request({
+    client: 'subscriberClient',
+    method: 'streamingPull',
+    returnFn: true,
+    reqOpts: {
+      highWaterMark: self.flowControl.highWaterMark
+    }
+  }, function(err, requestFn) {
     if (err) {
-      callback(err, resp);
+      self.emit('error', err);
       return;
     }
 
-    self.closed = true;
-    self.removeAllListeners();
+    self.connection = requestFn();
+
+    self.connection.on('error', function(err) {
```

Comment:
The way you handled users calling ".close()" makes sense to me. I would make sure "close" isn't triggered when there's an "error" as well, and your handling of one doesn't interfere with the other.